src/worker/worker.c

```diff
@@ -237,6 +240,27 @@ static int finish_job(child_process *cp, int reason)
 }
 
 /*
+ * Get the parent PID from a PID
+ */
+static void get_process_parent_id(const pid_t pid, pid_t * ppid) {
+	char buffer[BUFSIZ], *s_ppid;
+	FILE *fp;
+	sprintf(buffer, "/proc/%d/stat", pid);
+	fp = fopen(buffer, "r");
+	if (fp) {
+		size_t size = fread(buffer, sizeof (char), sizeof (buffer), fp);
+		if (size > 0) {
+			strtok(buffer, " "); // (1) pid  %d
+			strtok(NULL, " "); // (2) comm  %s
+			strtok(NULL, " "); // (3) state  %c
+			s_ppid = strtok(NULL, " "); // (4) ppid  %d
+			*ppid = atoi(s_ppid);
+		}
+		fclose(fp);
+	}
+}
```

Comment:
I refactored this one, adding a return value and checking in almost all places where something can go wrong. What do you think?
```c
static int get_process_parent_id(const pid_t pid, pid_t * ppid) {
        char buffer[BUFSIZ], *s_ppid;
        int errreading;
        FILE *fp;

        sprintf(buffer, "/proc/%d/stat", pid);
        fp = fopen(buffer, "r");
        if (!fp)
                return errno;

        size_t size = fread(buffer, sizeof (char), sizeof (buffer), fp);
        errreading = ferror(fp);
        if (fclose(fp) != 0)
                return errno;

        if (errreading != 0)
                return errreading;

        if (size == 0)
                return EBADF;

        strtok(buffer, " "); // (1) pid  %d
        strtok(NULL, " "); // (2) comm  %s
        strtok(NULL, " "); // (3) state  %c

        // (4) ppid  %d
        if ( (s_ppid = strtok(NULL, " ")) == NULL)
                return EBADF;

        *ppid = atoi(s_ppid);
        return 0;
}
```

Reply:
Any error checking for this line `*ppid = atoi(s_ppid);` ?

Reply:
I think checking `fclose(fp)` after checking `if (!fp)` isn't necessary. We'r checking a lot of os(operating system) system calls in this function.

Reply:
Regarding the `atoi` call: that is exactly why I commented that i validated *almost* all error possibilities, I would use `strtonum` instead. Now, related to your last comment, yeah, there are *a lot* of places where things can go wrong, to be sure or not that everything went well depends on how faithful you are.

Reply:
Ok. Would like to know if anyone else has any more comments on this function code?

Reply:
Done!